src/resolve/resolved-dns-synthesize.c

```diff
@@ -545,8 +542,11 @@ int dns_synthesize_answer(
                         if (u < 0)
                                 return log_error_errno(u, "Failed to synthesize local stub hostname PTR PR: %m");
 
-                        if (v == 0 && w == 0 && u == 0) /* This IP address is neither a local one, nor a gateway, nor a stub address */
+                        if (v == 0 && w == 0 && u == 0) {/* This IP address is neither a local one, nor a gateway, nor a stub address */
+                                if (dns_name_endswith(name, "127.in-addr.arpa") > 0)
+                                        nxdomain = true;
                                 continue;
+                        }
```

Comment:
I'd prefer if we'd actually generate a hostname for these addresses, for example `_47.localhost` or so, for 127.0.0.47.

Labelling is a *good* thing I am sure.

Reply:
I am not so sure. You would have to create also forward mapping from ``_47.localhost`` to resolve it to the correct address. And handling other records on such name to generate just NOERROR instead of NXDOMAIN. I think that would not be necessary for most of people. Ordinary users would usually not see these names. Advanced admins and developers can handle IP addresses directly. Just do not hide details from them. Such additional logic is unnecessary complicated in my opinion.

But *please* forget about leading underscore. There is absolutely no reason to use that. That would be just an ordinary hostname. If you are thinking about improvements, maybe consider adding SOA record with systemd indication to NXDOMAIN responses. Might indicate localhost zone ownership by systemd-resolved.

Reply:
i am not convinced that using literal decimal numbers as dns labels is such a good idea.

Reply:
I am quite sure just digits in a label are handled quite well today. It used to be mandatory to start hostname with a letter, but that is no longer required. If there were any doubt for backward compatibility, choosing ``_`` as its prefix seems to be much worse than leaving just digits. Definitely not improving.

Reply:
> I'd prefer if we'd actually generate a hostname for these addresses, for example `_47.localhost` or so, for 127.0.0.47.

That should at least become  `_0_0_47.localhost` or so in that case then :wink: 
Anyhow be it an underscore in front or a `h` etc doesn't make a difference imho, bottom line you guys should agree on is the title :wink: 

> I think that would not be necessary for most of people. Ordinary users would usually not see these names. 

Either don't do it, or do it properly, because you don't know if the end user falls in that generalization category.
_(that end-user might as well be a script)_

If i had to vote i would also vote like "Labelling is a good thing I am sure."

> You would have to create also forward mapping

Well if you want to play as a DNS server you should play it as it should and do it the proper way right?

Reply:
I maintain bind9 and unbound servers, which I consider proper enough. All of them have just entry for localhost and reverse entry for localhost and nothing else. I am quite sure they both are proper DNS servers, unlike systemd-resolved.

I have asked on DNS-OARC chat. [RFC 1123](https://www.rfc-editor.org/rfc/rfc1123.html#section-2) clearly specifies hostname can start with letter or a digit. But it does not say anything about other characters. I am not aware of any later specifications, which would define it in a different way. If course except names encoded in punnycode, in IDN 2003 or 2008 format.

Reply:
Original specification in [RFC 952](https://www.rfc-editor.org/rfc/rfc952.html) required hostname to start with a letter and not end with hyphen. Only [a-z0-9-] characters were allowed after the first letter.

Reply:
@poettering 
> i am not convinced that using literal decimal numbers as dns labels is such a good idea, hence prefix them with something

It's perfectly fine, see RFC 2317, and reverse records in general:
```sh
$ dig +short PTR 71.157.214.85.in-addr.arpa.
gardel.0pointer.net.
```

@TriMoon 
> Please don't code how things are "handled" at any moment in time, but stick to the specifications of that technology else you will end-up being the odd-duck at some point in time.

This is not just handled well, it is also expressly compliant with the specs. Unlike `_`, whose handling is largely unknown on the end-user side of systems (backend things like DNS-internal signaling notwithstanding).

Taking you up with your point of being the odd-duck, it would seem wise to avoid the underscore kind of labeling.

Apart from this spec-level argument, I share @pemensik's perspective that people who would read `47.localhost` would have not benefit from seeing that over seeing `127.0.0.47` directly. It's entirely unclear why
> Labelling is a *good* thing
in this case. In any case, the additional complexity in coding and human processing does not speak in favor of it. Are there any additional arguments for why some consider it good?

Reply:
> Are there any additional arguments for why some consider it good?

Pure clarity for the human, just like comments in code that don't affect the actual code logic... :wink: 
Sometimes that clarity needs more complexity in coding to make the output for the human more clear, just like a GUI vs a text terminal. :wink: 
Imagine the first GUI coders would have said "Why bother so much coding?" :rofl:

Reply:
If there are spare resources for doing extra work, I would like to request time spent on designing fix for issue #19227, instead of localhost special name generation. Thanks!